Create `dart run` page (#3154)
dart-lang/site-www: src/tools/dart-run.md

Here's an example of creating a new app and running it:

```terminal
$ dart create myapp
$ cd myapp
$ dart run
```

## Running a script in your package's bin directory

This is the simplest use case.

To run a script in a package named `foo`,
which contains `bar.dart` in the `bin` directory,
run the following command from the root of the package,
optionally passing zero or more arguments:

```terminal
$ dart run foo:bar [arg1 arg2 ...]
```

If your package and script file in `bin` have the same name
(e.g. `/foo/bin/foo.dart`),
then you can use the following shorthand:

```terminal
$ dart run foo [arg1 arg2]
```

## Running a script in another directory in your package

To run a script inside a directory other than the top-level
bin directory (but within the package),
prepend the relative path to the name of the script.

For example, to run `foo.dart` in the `example/sub` directory:

```terminal
$ dart run example/sub/foo.dart [arg1 arg2 ...]
```

## Running a script in a dependency

To run a script from the `bin` directory of a package that you depend on
in the pubspec, specify the package name.
For example, to run `bar.dart` in the `baz` package:

```terminal

> $ dart run baz:bar [arg1 arg2 ...]

Reviewer: Here it might be nice to illustrate the shorthand again... This is usually where it's useful.

`dart test` is actually a shorthand for `dart run test` which is a shorthand for `dart run test:test`.

And this is actually the case where you're running a script in a dependency.

---
title: dart run
deprogramion: Use dart run to run a Dart program in your package.
---

The `dart run` command supports running a Dart program located in a file,
or a package, or in one of its dependencies, from the command line.

To run an executable when you are not currently inside a package,
use the [pub global](/tools/pub/cmd/pub-global) command.

The `dart run` command combines what was previously exposed as `pub run`
and as the [Dart VM](/tools/dart-vm).

{% include tools/dart-tool-note.md %}

Here's an example of creating a new app and running it:

```terminal
$ dart create myapp
$ cd myapp
$ dart run
```

## Running a program in a Dart file

This is the simplest form, where you pass a relative path to a `.dart` file.

> To run a program in the file `internal/debug.dart`,

Reviewer: nit: `test/debug.dart` might be better... `internal/` is not a convention.

Author: done

is considered a package. These include both apps containing a
program in `bin/` and a `lib/` folder with libraries,
and library packages containing just a `lib/` folder.

### Running a program from a package dependency

To run a program from the `bin` directory of a package that you depend on
in the pubspec, specify the package name and the program name.
For example, to run `bin/bar.dart` in the `bar` package:

```terminal
$ dart run bar [arg1 arg2 ...]
```

If the program name doesn't match the package name,
use the form `<package name>:<program name>`. For example,
to run the program `bin/baz.dart` in `package:bar` use:

```terminal
$ dart run bar:baz [arg1 arg2 ...]
```

You can only run programs out of another package's `bin` directory.
All other directories are private.

### Running a program in the current package

When the current directory is the root folder of a package,
you can use the following short form,
which omits the package name and just specifies the program name,
to run `bin/baz.dart`:

```terminal
$ dart run :baz [arg1 arg2 ...]
```

To run the main program -- the one which has the same name as the package --
you can use these short forms,

> which will run `<current directory>/bin/<current directory>.dart`:

Reviewer: Actually, this is only if `<current directory> == "name" from pubspec.yaml`.

Name the of the current package is not the name of the current directory, it's the `name:` property from `./pubspec.yaml`.

Author: clarified

---
title: dart run

> deprogramion: Use dart run to run a Dart program in your package.

Reviewer 2: I'm assuming you meant `description` here and from the content it seems we want to describe running programs in more than just "your" package.

```suggestion
description: Use dart run to run a Dart program.
```

---

The `dart run` command supports running a Dart program located in a file,

> or a package, or in one of its dependencies, from the command line.

Reviewer 2: Perhaps the first "or" isn't needed here.

```suggestion
a package, or in one of its dependencies, from the command line.
```

is considered a package. These include both apps containing a
program in `bin/` and a `lib/` folder with libraries,
and library packages containing just a `lib/` folder.

### Running a program from a package dependency

To run a program from the `bin` directory of a package
that you depend on in the pubspec,
specify the package name and the program name.
For example, to run `bin/bar.dart` in the `bar` package:

```terminal
$ dart run bar [arg1 arg2 ...]
```

If the program name doesn't match the package name,
use the form `<package name>:<program name>`. For example,
to run the program `bin/baz.dart` in `package:bar` use:

```terminal
$ dart run bar:baz [arg1 arg2 ...]
```

You can only run programs out of another package's `bin` directory.
All other directories are private.

### Running a program in the current package

When the current directory matches the package name
(i.e. the `name:` property in the pubspec),
you can use the following short form,
which just specifies the program name:

```terminal
$ dart run :baz [arg1 arg2 ...]
```

To run the main program
-- the one which has the same name as the package --
use these short forms:

```terminal
$ dart run

> $ dart run [-- arg1 arg2 ...]

Reviewer 2: Is it necessary to outline both here? To me, these really seems like the same "short form".

```

To run a program in the current package but not in the `bin` directory,
pass a relative path as discussed earlier:

```terminal
$ dart run tool/debug.dart [arg1 arg2 ...]
```

> ## Running with debugging

Reviewer 2: Since we're including some information about the debugging flags, should we mention the ability to enable experiments and perhaps link to https://dart.dev/tools/experiment-flags?

For example, to run `foo.dart` in the `example/sub` directory:

```terminal

> $ dart run example/sub/foo arg1 arg2

Reviewer: This won't work, `dart run example/sub/foo.dart` will work

> For example, to run `bar.dart` in the foo package:

Reviewer: I would rephrase this as something like: runs `bin/bar.dart` from `package:foo`.

Reviewer: TODO: we should also highlight the shorthand `dart run foo` is equivalent to `dart run foo:foo`, which means it runs `bin/foo.dart` from `package:foo`.
